### PR TITLE
Fix domain listing per-domain parent lookup N+1 (slow login at scale)

### LINKED ISSUE #29379: Feature: keep the Domains experience fast & usable for large-scale (tens of thousands of domains)
### Is your feature request related to a problem? Please describe.

Organizations that sync their domain taxonomy from an external system (e.g. via an ETL job) can end up with **tens of thousands of domains nested several levels deep** (we hit this at ~20k domains, depth ~3, on K8s + OpenSearch + Postgres). At that scale the Domains experience degrades:

- The post-login landing page (Domains widget + asset counts) can take up to ~1 minute and time out.
- The global domain selector / tree and the Domains listing page feel sluggish to open and browse.
- Child counts and hierarchy navigation are slow.

For a data steward, this makes one of the core governance surfaces frustrating to use exactly when the catalog is large enough to need it most.

### Describe the solution you'd like

A Domains experience that stays responsive regardless of catalog size:

- The global **domain selector / tree** opens and paginates instantly, lazily loading children, and doesn't re-fetch on every open within a session.
- The **Domains listing page** and **hierarchy** load quickly, with accurate nested child counts.
- **Post-login landing** (domain widget + asset counts) never blocks or times out.
- The underlying APIs (`GET /domains`, `GET /domains/hierarchy`, `GET /domains/assets/counts`) return quickly at p95 even at 20k+ domains, on both cold and warm caches.

This is the user-facing outcome; the engineering/scale work to get there (a materialized, cross-pod hierarchy cache, subtree-aware invalidation, frontend caching for the tree, and generalizing the batching to other hierarchical entities) is tracked in #29332. The cold-load groundwork has already landed in PR #29247.

### Describe alternatives you've considered

- **Manual pagination with small page sizes** (the current workaround) — keeps requests from timing out but makes browsing tedious and doesn't help the landing page.
- **Capping domain depth / count** — not acceptable for organizations that legitimately have large, deep taxonomies.
- **Algorithmic fixes only** (delivered in #29247) — necessary and sufficient for cold loads, but the warm / high-traffic path and multi-replica deployments still benefit from caching.

### Additional context

- Reported scale: ~20,000 domains, ~3 levels deep, synced via an external ETL, running on Kubernetes with OpenSearch and CloudNativePG (Postgres).
- Related: #29332 (engineering tracking for the scale/caching work), PR #29247 (delivered cold-load N+1 fixes + cache-aware parent + post-login widget caching), #26068 (similar N+1 at scale for tables).

### Pre-submission checklist

- [x] I searched for duplicate feature requests.

### PR DESCRIPTION
**Pull request by harshach**
### Describe your changes:

Related to #29379

On catalogs with ~20k nested domains, logins / cold page loads timed out because `DomainRepository.setInheritedFields` resolved every domain's parent unconditionally, so listing all domains (the post-login `/domains/assets/counts` widget call) issued ~20k relationship + parent queries. This reuses the base batched inheritance path (gated by `requiresParentForInheritance`, with `fetchInheritableRelationships` overridden to also load experts on parent domains), batches `childrenCount` into one scan per parent, lightens `/domains/assets/counts` to an FQN-only query, routes single-domain parent resolution through the cache-aware container cache and warms that cache from the batch path, and caches the DomainsWidget calls via React Query so repeat landing-page visits don't refetch.

#
### Type of change:
- [x] Bug fix
- [x] Improvement

#
#### Use cases covered
- Listing/loading all domains at ~20k scale no longer issues a parent lookup per domain.
- A subdomain with no explicit owners/experts still inherits them from its ancestor.
- `childrenCount` (nested) and `/domains/assets/counts` return the same values as before.

#### Unit tests
- `DomainRepositoryListingTest` — query-count RED→GREEN: listing without owners/experts issues no per-domain parent lookup; `childrenCount` collapses to one scan per parent (no per-domain `COUNT`).

#### Backend integration tests
- `DomainResourceIT#test_subDomainInheritsOwnersExpertsAndChildrenCount` — depth-2 owners/experts inheritance, nested `childrenCount`, and the bulk list path.

#### Playwright (UI) tests
- [x] Not applicable — covered by Jest (`DomainsWidget.test.tsx`: no refetch within `staleTime`).

#
### UI screen recording / screenshots:

Not applicable (caching only, no visual change).

#
### Checklist:
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests (unit / integration) and listed them above.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details open><summary><h3>Greptile Summary</h3></summary>

This PR eliminates an N+1 query regression in domain listing that caused login timeouts on catalogs with ~20k nested domains. It replaces per-domain parent resolution and per-domain `COUNT(*)` queries with batched equivalents, lightens the `/domains/assets/counts` call to an FQN-only scan, and adds a 5-minute React Query cache on the DomainsWidget to prevent redundant post-login fetches.

- **Backend (DomainRepository):** Removes the custom `setFieldsInBulk` override and hooks into the base batched inheritance path via `requiresParentForInheritance`/`fetchInheritableRelationships`/`applyInheritance`. `childrenCount` is now computed with a single selective fqnHash prefix scan (or one full scan for cross-root pages) via the new `DomainHierarchyHashes` utility, replacing N per-domain `COUNT(*)` calls. The batch parent fetch also warms the container cache so single-domain reads after a list call hit the cache instead of the DB.
- **Frontend (DomainsWidget):** Wraps `searchQuery` and `getAllDomainsWithAssetsCount` in `queryClient.fetchQuery` with a 5-minute `staleTime`, so repeat landing-page visits within a session serve from cache. Sort changes explicitly invalidate the affected query key to force a fresh fetch.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the batching logic is correct, algorithm is well-tested, and the change is strictly additive in DB query behavior.

The core DomainHierarchyHashes math is pure and verified by exhaustive unit tests including edge cases (ancestor-descendant pairs on the same page, multi-root fallback, empty pages). The nestedDescendantCounts path correctly mimics the old countNestedDomains FQN-hash prefix semantics. The batchFetchParents refactor resolves parent refs with a two-query batch and safely null-guards the container-cache warming. The buildDomainAssetCounts FQN-only query is a straightforward table scan replacing a full entity hydration. The UI cache layer uses standard React Query patterns with bounded stale time and explicit invalidation on sort change.

No files require special attention.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java | Core N+1 fix: replaces per-domain parent resolution and per-domain COUNT(*) with batched equivalents; delegates inheritance to the base class batched path via requiresParentForInheritance/applyInheritance overrides; adds container-cache warming in batchFetchParents; and streamlines buildDomainAssetCounts to an FQN-only query. |
| openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainHierarchyHashes.java | New utility class implementing pure fqnHash hierarchy math (longestCommonAncestor, strictAncestorHashes, countDescendantsByAncestor); no DB or entity dependencies; well-tested by DomainHierarchyHashesTest. |
| openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java | Adds three new DAO methods on DomainDAO: listFqnHashesByPrefix (selective prefix scan), listAllFqnHashes (full hash scan), and listAllFqns (dual Postgres/MySQL JSON extraction ordered by fqnHash). |
| openmetadata-service/src/test/java/org/openmetadata/service/jdbi3/DomainRepositoryListingTest.java | New unit test guarding query-count behavior: verifies no per-domain parent lookup without inheritable fields, that childrenCount batches to one scan per parent, and that multi-root pages fall back to a single full scan. |
| openmetadata-service/src/test/java/org/openmetadata/service/jdbi3/DomainHierarchyHashesTest.java | Comprehensive unit tests for DomainHierarchyHashes covering edge cases: empty input, single hash, siblings, ancestor-descendant pairs, shared character prefix with different segments, multi-depth attribution. |
| openmetadata-integration-tests/src/test/java/org/openmetadata/it/tests/DomainResourceIT.java | Adds integration test verifying depth-2 owner/expert inheritance, nested childrenCount, and that the bulk list path also inherits owners on subdomains. |
| openmetadata-ui/src/main/resources/ui/src/components/MyData/Widgets/DomainsWidget/DomainsWidget.tsx | Wraps domain widget fetches in queryClient.fetchQuery with a 5-minute staleTime; sort changes invalidate only the affected search key, leaving the asset-count cache intact. |
| openmetadata-ui/src/main/resources/ui/src/rest/queries/domainQuery.ts | Adds DOMAIN_WIDGET_STALE_TIME constant and two typed query key factories (domainAssetsCountQueryKey, domainWidgetSearchQueryKey) for the DomainsWidget cache. |
| openmetadata-ui/src/main/resources/ui/src/components/MyData/Widgets/DomainsWidget/DomainsWidget.test.tsx | Adds a cache test that verifies getAllDomainsWithAssetsCount is called only once across mount→unmount→remount within the staleTime window; clears queryClient in beforeEach to prevent inter-test pollution. |

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Client
    participant DomainResource
    participant DomainRepository
    participant DB

    note over Client,DB: Before (N+1 per-domain queries for 20k domains)
    Client->>DomainResource: "GET /domains?fields=owners,experts,childrenCount"
    DomainResource->>DomainRepository: setFieldsInBulk(domains)
    loop for each domain (N queries)
        DomainRepository->>DB: getFromEntityRef (parent lookup)
        DomainRepository->>DB: Entity.getEntity (load parent with owners+experts)
        DomainRepository->>DB: "countNestedDomains COUNT(*) per domain"
    end
    DomainRepository-->>Client: response

    note over Client,DB: After (batched — 3 queries for the whole page)
    Client->>DomainResource: "GET /domains?fields=owners,experts,childrenCount"
    DomainResource->>DomainRepository: setFieldsInBulk(domains)
    DomainRepository->>DB: findFromBatch (all parents in one query)
    DomainRepository->>DB: getEntityReferencesByIds (resolve parent refs, warm container cache)
    DomainRepository->>DB: listFqnHashesByPrefix OR listAllFqnHashes (one scan for childrenCount)
    note right of DomainRepository: in-memory countDescendantsByAncestor attributes descendants to each page domain
    DomainRepository-->>Client: response (owners/experts inherited from pre-loaded parents)
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Client
    participant DomainResource
    participant DomainRepository
    participant DB

    note over Client,DB: Before (N+1 per-domain queries for 20k domains)
    Client->>DomainResource: "GET /domains?fields=owners,experts,childrenCount"
    DomainResource->>DomainRepository: setFieldsInBulk(domains)
    loop for each domain (N queries)
        DomainRepository->>DB: getFromEntityRef (parent lookup)
        DomainRepository->>DB: Entity.getEntity (load parent with owners+experts)
        DomainRepository->>DB: "countNestedDomains COUNT(*) per domain"
    end
    DomainRepository-->>Client: response

    note over Client,DB: After (batched — 3 queries for the whole page)
    Client->>DomainResource: "GET /domains?fields=owners,experts,childrenCount"
    DomainResource->>DomainRepository: setFieldsInBulk(domains)
    DomainRepository->>DB: findFromBatch (all parents in one query)
    DomainRepository->>DB: getEntityReferencesByIds (resolve parent refs, warm container cache)
    DomainRepository->>DB: listFqnHashesByPrefix OR listAllFqnHashes (one scan for childrenCount)
    note right of DomainRepository: in-memory countDescendantsByAncestor attributes descendants to each page domain
    DomainRepository-->>Client: response (owners/experts inherited from pre-loaded parents)
```

</a>
</details>

<sub>Reviews (9): Last reviewed commit: ["Merge branch &#39;main&#39; into harshach/domain..."](https://github.com/open-metadata/openmetadata/commit/c4c4223a5dafbca4576cb1cfaac6af8749e1840b) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38779164)</sub>

### CODE REVIEW

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java [hunk 1]
```diff
@@ -151,14 +154,62 @@ private void fetchAndSetExperts(List<Domain> domains, Fields fields) {
...
     for (Domain entity : entities) {
-      int count = daoCollection.domainDAO().countNestedDomains(entity.getFullyQualifiedName());
-      entity.setChildrenCount(count);
+      String childPrefix = hashById.get(entity.getId()) + Entity.SEPARATOR;
+      int count = 0;
+      for (String descendantHash : descendantHashes) {
+        if (descendantHash.startsWith(childPrefix)) {
+          count++;
+        }
+      }
```
[the remark this answers is not in the document]
Comment (harshach) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java]: Fixed in 081735252c. `batchCountNestedDomains` now does a single O(descendants × depth) pass: it walks each descendant hash's prefix boundaries and attributes it to any page-domain hash that matches, instead of re-scanning the full descendant list per entity. A flat list of all domains with `childrenCount` is now linear, not quadratic.

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java [hunk 2]
```diff
@@ -151,14 +154,62 @@ private void fetchAndSetExperts(List<Domain> domains, Fields fields) {
...
+      if (lastSeparator > 0) {
+        parentPrefixes.add(childHash.substring(0, lastSeparator));
+      } else {
+        hasRootDomain = true;
+      }
+    }
+    List<String> descendantHashes;
+    if (hasRootDomain) {
+      descendantHashes = daoCollection.domainDAO().listAllFqnHashes();
+    } else {
```
[the remark this answers is not in the document]
Comment (harshach) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java]: Fixed in 081735252c. `fetchDescendantHashes` now issues **at most one** query: a selective `LIKE prefix.%` scan only when the whole page shares a single parent (the tree-expand case), otherwise a single index-only full scan of `fqnHash` — never a query-per-parent fan-out. Added `DomainRepositoryListingTest#childrenCount_forPageSpanningMultipleParents_usesOneFullScan` covering a page spanning multiple parents.

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java [hunk 3]
```diff
@@ -151,14 +154,62 @@ private void fetchAndSetExperts(List<Domain> domains, Fields fields) {
...
+      descendantHashes = new ArrayList<>();
+      for (String parentPrefix : parentPrefixes) {
+        descendantHashes.addAll(
+            daoCollection.domainDAO().listFqnHashesByPrefix(parentPrefix + ".%"));
+      }
     }
+    return descendantHashes;
   }
```
[the remark this answers is not in the document]
Comment (harshach) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java]: Fixed in 081735252c — collapsed to at most one query: a single selective prefix scan for a shared-parent page, else one index-only full scan. So a mixed-hierarchy / multi-parent page (incl. the `A.B` + `A.B.C` case) no longer fans out to k queries, and with a single result set there is no cross-query overlap to dedup. Added a multi-parent test.

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java [hunk 4]
```diff
@@ -720,6 +786,13 @@ protected EntityReference getParentReference(Domain entity) {
     return entity.getParent();
   }
 
+  private EntityReference resolveParentRef(Domain entity) {
+    // Use the cache-aware CONTAINS container fast-path (fromEntityType == null) so warm reads hit
+    // the relationship-container cache when distributed caching is enabled. A domain's only
+    // CONTAINS parent is its parent domain, so a null fromEntityType is unambiguous here.
+    return getFromEntityRef(entity.getId(), DOMAIN, Relationship.CONTAINS, null, false);
+  }
```
[the remark this answers is not in the document]
Comment (harshach) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java:798]: Aligned in 081735252c — `batchFetchParents` now resolves parent refs with `Include.ALL` to match the single-read `getFromEntityRef` path that reads/writes the same container cache. Domains don't support soft delete (`supportsSoftDelete=false`), so ALL == NON_DELETED today; added a comment documenting the assumption.

[the remark this answers is not in the document]
Comment (harshach) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java:798]: Intentional, and it has to stay `null`: the cache-aware container fast-path in the 5-arg `getFromEntityRef` only engages for a type-agnostic CONTAINS lookup (`fromEntityType == null`). Passing `DOMAIN` would be clearer but would bypass the relationship-container cache entirely and defeat the warm-read optimization this PR adds. A domain's only CONTAINS parent is its parent domain, so the lookup is unambiguous today; added a comment (081735252c) noting that if another entity type is ever allowed to CONTAIN a domain, the container-cache contract itself — not just this call — would need revisiting.

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java [hunk 5]
```diff
@@ -151,14 +153,72 @@ private void fetchAndSetExperts(List<Domain> domains, Fields fields) {
...
     }
+    // A single shared parent (the tree-expand case) → one selective prefix scan. A root page or a
+    // flat page spanning multiple parents → one index-only full scan, never a query-per-parent
+    // fan-out (which would re-introduce an N pattern for mixed-hierarchy pages).
+    List<String> descendantHashes;
+    if (!hasRootDomain && parentPrefixes.size() == 1) {
+      descendantHashes =
+          daoCollection.domainDAO().listFqnHashesByPrefix(parentPrefixes.iterator().next() + ".%");
+    } else {
+      descendantHashes = daoCollection.domainDAO().listAllFqnHashes();
```
Comment (sonika-shah) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java]: **perf (non-blocking):** this `else` branch calls `listAllFqnHashes()` (`SELECT fqnHash FROM domain_entity`, no filter) — a **full table scan of every domain** (~20k rows) pulled into memory.

It's hit whenever the page is **not** a single set of same-parent siblings — i.e. it includes a root domain or spans >1 parent. So a flat domain list requested with `?fields=childrenCount` that spans multiple parents does a full scan **on every page** → listing all domains ≈ (total / pageSize) full scans, each re-materializing all fqnHashes. (The old per-domain `COUNT(*)` held nothing in memory; this trades that for the scan.)

Not a blocker — still far better than the old `COUNT(*)` N+1, and the sibling tree-expand path (the `if` branch, `listFqnHashesByPrefix`) is a selective scan and unaffected. Fix: memoize this scan for the duration of a request (the page set is immutable within a request) so repeated pages reuse one read.

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java [hunk 6]
```diff
@@ -266,10 +342,9 @@ public Map<String, Integer> getAllDomainsWithAssetsCount() {
       return new HashMap<>();
     }
 
-    List<Domain> allDomains = listAll(getFields("fullyQualifiedName"), new ListFilter(null));
+    List<String> allFqns = daoCollection.domainDAO().listAllFqns();
```
Comment (sonika-shah) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java:337]: **nit (non-blocking):** `listAllFqns()` has no `ORDER BY`, so this `LinkedHashMap` — and the `/domains/assets/counts` response — is now in DB-arbitrary order (old `listAll` was ordered). Harmless today (UI re-sorts). Add `ORDER BY fqnHash` to the query if any consumer relies on order.

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java (hunk 5 above)
Comment (harshach) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java]: Addressed in 52b0582cd3. `fetchDescendantHashes` now scans under the **longest common ancestor** of the whole page in a single query: siblings collapse to their parent, and a multi-parent page collapses to its shared top-level ancestor — a selective index range, not a full table scan. Only a page whose domains span *unrelated* top-level domains (no shared ancestor) falls back to the full `listAllFqnHashes()` scan, which is unavoidable in one query.

On the request-memo idea: each paginated page is a separate HTTP request doing exactly one scan, so there's nothing to reuse within a request — the common-ancestor collapse is what actually shrinks the per-page cost (a flat list paginated in fqnHash order keeps consecutive pages under one root → selective). Added tests for the common-ancestor selective scan and the unrelated-roots full-scan fallback.

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java (hunk 6 above)
Comment (harshach) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java:337]: Done in 52b0582cd3 — added `ORDER BY fqnHash` to `listAllFqns()` (both MySQL and Postgres variants), so `GET /domains/assets/counts` is deterministically ordered by fqnHash again. fqnHash is the unique index, so it stays an index-ordered read (no filesort).

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java [hunk 7]
```diff
@@ -151,14 +153,97 @@ private void fetchAndSetExperts(List<Domain> domains, Fields fields) {
...
+    }
+    String result;
+    if (i == a.length() && (i == b.length() || b.startsWith(Entity.SEPARATOR, i))) {
+      result = a; // a == b, or a is an ancestor of b
+    } else if (i == b.length() && a.startsWith(Entity.SEPARATOR, i)) {
+      result = b; // b is an ancestor of a
+    } else {
+      // Common chars stop mid-segment — truncate to the last whole segment they share.
+      int lastSeparator = a.lastIndexOf(Entity.SEPARATOR, i - 1);
+      result = lastSeparator > 0 ? a.substring(0, lastSeparator) : "";
```
[the remark this answers is not in the document]
Comment (harshach) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/DomainRepository.java]: Addressed in dc8cc64625 — the whole char-index helper (`commonAncestorOf` with `a`/`b`/`i`/`max`) is gone. The fqnHash math now lives in a pure, `final`, directly unit-tested helper `DomainHierarchyHashes` that operates on **segment lists** instead of character offsets, with prose names: `longestCommonAncestor`, `strictAncestorHashes`, `countDescendantsByAncestor`, and `longestCommonPrefix(left, right)` using `sharedDepth`/`comparableDepth`. `DomainRepository` is now a thin orchestrator (`nestedDescendantCounts`, `fetchCandidateDescendantHashes`). 14 new direct unit tests cover the helper.

**Review by karanh37 — DISMISSED** [review]

**Review by sonika-shah — DISMISSED** [review]
